agent/acs/handler/attach_eni_handler.go

```diff
@@ -118,14 +129,36 @@ func (attachENIHandler *attachENIHandler) handleSingleMessage(message *ecsacs.At
 // addENIAttachmentToState adds the eni info to the state
 func (handler *attachENIHandler) addENIAttachmentToState(message *ecsacs.AttachTaskNetworkInterfacesMessage) {
 	attachmentArn := aws.StringValue(message.ElasticNetworkInterfaces[0].AttachmentArn)
+	eniMac := aws.StringValue(message.ElasticNetworkInterfaces[0].MacAddress)
+	// Stop tracking the eni attachment after timeout
+	eniAckTimeout := time.Duration(aws.Int64Value(message.WaitTimeoutMs)) * time.Millisecond
+	_time := ttime.DefaultTime{}
+	ackTimeoutTimer := _time.AfterFunc(eniAckTimeout, handler.ackTimeoutHandler(eniMac))
 	seelog.Info("Adding eni info to state, eni: %s", attachmentArn)
-	handler.state.AddENIAttachment(
-		&api.ENIAttachment{
-			TaskArn:          aws.StringValue(message.TaskArn),
-			AttachmentArn:    attachmentArn,
-			AttachStatusSent: false,
-			MacAddress:       aws.StringValue(message.ElasticNetworkInterfaces[0].MacAddress),
-		})
+
+	eniAttachment := &api.ENIAttachment{
+		TaskArn:          aws.StringValue(message.TaskArn),
+		AttachmentArn:    attachmentArn,
+		AttachStatusSent: false,
+		MacAddress:       eniMac,
+		AckTimer:         ackTimeoutTimer,
+	}
+	handler.state.AddENIAttachment(eniAttachment)
+}
+
+// ackTimeoutHandler returns the function to remove ENI attachment from agent state
+// after the ENI ack timeout
+func (handler *attachENIHandler) ackTimeoutHandler(mac string) func() {
+	return func() {
```

Comment:
My comment in the previous revision was to avoid using anonymous functions as they lead to code that cannot be tested properly. In this instance, it's very hard to write unit tests for `ackTimeoutHandler` method itself. Can you refactor this instead to be a named method?

```
type ackTimeoutHandler struct {
 mac string
 state dockerstate.TaskEngineState
}

func (handler ackTimeoutHandler) handle() {
 eniAttachment, ok := handler.state.ENIByMac(mac)
 if !ok {
  seelog.Warnf("ENI attachment not managed by agent for ENI with MAC address: %s", mac)
  return
 }
 
 if !eniAttachment.GetSentStatus() {
  handler.state.RemoveENIAttachment(mac)
 }
}

// When you create the attachment, you do:
handler := ackTimeoutHandler{mac:eniMac,...}
ackTimeoutTimer := time.AfterFunc(eniAckTimeout, handler)
eniAttachment := &api.ENIAttachment{
 ...  
 AckTimer: ackTimeoutTimer,
}
```